warehouse/oidc/services.py

```diff
@@ -219,6 +220,23 @@ def _get_key_for_token(self, token):
         unverified_header = jwt.get_unverified_header(token)
         return self._get_key(unverified_header["kid"])
 
+    def token_identifier_exists(self, jti: str) -> bool:
+        """
+        Check if a JWT Token Identifier has already been used.
+        """
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            return bool(r.exists(jti))
+
+    def store_jwt_identifier(self, jti: str, expiration: int) -> None:
+        """
+        Store the JTI with its expiration date if the key does not exist.
+        """
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            # Defensive: to prevent races, we expire the JTI slightly after
+            # the token expiration date. Thus, the lock will not be
+            # released before the token invalidation.
+            r.set(jti, exat=expiration + 5, value="placeholder", nx=True)
```

Comment:
I _think_ we need to do some namespacing/key distinction here: right now two different IdPs could in theory (not likely in practice) send the same `jti`, which would result in a false positive. 

Similar to how the JWKS set is stored under `/warehouse/oidc/jwks/{self.publisher}`, I think we probably want to do `/warehouse/oidc/{self.publisher}/{jti}` here 🙁

Reply:
(Separately from `jti` confusion between IdPs, a compromised IdP could also use the non-namespaced implementation to set an arbitrary Redis key to an integer they control, which could be a security issue e.g. for ratelimiting and session tracking state.)